### PR TITLE
NucleicAcidSearchEngine improvements

### PR DESCRIPTION
**Pull request by hendrikweisser**
- Allow negative values for parameter "precursor:isotopes" (see issue #4168)
- Take precursor intensities into account when estimating target RTs for label-free quantification (see parameter "lfq_out") - this improves feature detection using FeatureFinderMetaboIdent
- Add log messages with number of identified spectra (before/after FDR filtering)

### CODE REVIEW

File: src/utils/NucleicAcidSearchEngine.cpp [hunk 1]
```diff
@@ -846,21 +858,27 @@ class NucleicAcidSearchEngine :
...
   }
 
 
   void generateLFQInput_(IdentificationData& id_data, const String& out_file)
   {
-    // mapping: (oligo, adduct) -> charge -> RTs
-    map<pair<NASequence, String>, map<Int, vector<double>>> rt_info;
+    // mapping: (oligo, adduct) -> charge -> (precursor intensity, RT)
+    map<pair<NASequence, String>,
+        map<Int, vector<pair<double, double>>>> rt_info;
```
Comment (timosachsenberg) [inline src/utils/NucleicAcidSearchEngine.cpp]: Hard to read :) typedefs?

Comment (cbielow) [reply src/utils/NucleicAcidSearchEngine.cpp]: ehh.. maps within a map? Are you sure?
is not `map<a, map<b, c>>` better represented as map `< (a,b), c>` in terms of efficiency?
depends a bit on the use case I guess (do you want to count charges, for example?)

Comment (hendrikweisser) [reply src/utils/NucleicAcidSearchEngine.cpp]: @cbielow: I don't think efficiency is a concern at that point. The code only gets executed once per tool run. In `map<pair<a, b>, c>`, the `a`s are stored multiple times, which may make it more or less efficient.

**Comment by timosachsenberg** [comment]
rebuild jenkins

File: src/utils/NucleicAcidSearchEngine.cpp [hunk 2]
```diff
@@ -879,13 +899,20 @@ class NucleicAcidSearchEngine :
...
+        for (const auto& rt_pair : charge_pair.second)
+        {
+          weighted_rt += rt_pair.first * rt_pair.second;
+          total_weight += rt_pair.first;
+        }
+        rts.push_back(weighted_rt / total_weight);
       }
       tsv << name << ef << 0 << ListUtils::concatenate(charges, ",");
-      // @TODO: do something more sophisticated for the RT?
+      // overall target RT is median over all charge states:
```
Comment (timosachsenberg) [inline src/utils/NucleicAcidSearchEngine.cpp]: without seeing the full picture this sounds a bit dangerous

Comment (hendrikweisser) [reply src/utils/NucleicAcidSearchEngine.cpp]: My reasoning is that LC happens before ionisation, so in principle all charge states must have the same RT. I'm trying to alleviate problems from patchy MS2 sampling by combining data from multiple charge states.

Comment (timosachsenberg) [reply src/utils/NucleicAcidSearchEngine.cpp]: I don’t know much about that but I thought that this Depends on the adduct.

Comment (hendrikweisser) [reply src/utils/NucleicAcidSearchEngine.cpp]: Not sure if that's what you mean, but the RT calculations are done separately for every combination of oligonucleotide and adduct. So you can get different RTs for the same oligo with different adducts.

Comment (timosachsenberg) [reply src/utils/NucleicAcidSearchEngine.cpp]: so every oligo + adduct pair get 1 target RT. If several charge states of the same oligo + adduct pair exist then also the median of all RTs is returned?

Comment (hendrikweisser) [reply src/utils/NucleicAcidSearchEngine.cpp]: Yes. One final RT per oligo/adduct pair, which is calculated from all charge states of that oligo/adduct pair.

Comment (timosachsenberg) [reply src/utils/NucleicAcidSearchEngine.cpp]: Ok fine with me. I only see problems with e.g. deuterium as charge carrier but this sounds kind of artificial.

**Comment by timosachsenberg** [comment]
rebuild jenkins

**Comment by timosachsenberg** [comment]
rebuild jenkins

**Pull request merged**